xaitk_saliency/interfaces/vis_sal_detection.py

```diff
@@ -0,0 +1,75 @@
+import abc
+
+import numpy as np
+from smqtk_core import Plugfigurable
+
+
+class ImageDetectionSaliencyMapGenerator (Plugfigurable):
+    """
+    This interface proposes that implementations transform black-box image
+    object detection predictions into visual saliency heatmaps.
+    This should require externally-generated object detection predictions over
+    some image, along with predictions for perturbed images and the permutation
+    masks for those images as would be output from a
+    :class:`xaitk_saliency.interfaces.perturb_image.PerturbImage`
+    implementation.
+
+    Object detection representations used here would need to encapsulate
+    localization information (i.e. bounding box regions), class scores, and
+    objectness scores (if applicable to the detector, such as YOLOv3).
+    Object detections are converted into (4+1+nClasses) vectors (4 indices for
+    bounding box locations, 1 index for objectness, and nClasses indices for
+    different object classes).
+    """
+
+    @abc.abstractmethod
+    def generate(
+        self,
+        ref_dets: np.ndarray,
+        perturbed_dets: np.ndarray,
+        perturb_masks: np.ndarray,
+    ) -> np.ndarray:
+        """
+        Generate visual saliency heat-map matrices for each reference
+        detection, describing what visual information contributed to the
+        associated reference detection.
+
+        We expect input detections to come from a black-box source that outputs
+        our minimum requirements of a bounding-box, objectness and per-class
+        scores.
```

Comment:
Maybe add something like: "For architectures which do not produce objectness scores, a default value of 1.0 is used." This goes back to the question of how we plan to encapsulate a detection object.

Reply:
> This goes back to the question of how we plan to encapsulate a detection object.

Yeah, I think we have a number of options on that front. My thought there was to use some basic types in this first pass (i.e. numpy), and then with some additional time maybe we could define a "DetectedObjectSet" (maybe defined in smqtk-detectors) that exposes an access API for the required bits, enforcing acceptable defaults where applicable. I think that's a few steps ahead of where we currently are, though.

Reply:
I like your recommended phrasing, maybe I'll tweak that though to say that the objectness column should still be provided, but provide the guidance of using the 1.0 value in the absence of it being a natural output of some detector impl.

Reply:
Sounds good to me!